### ISSUE TITLE
On non-US keyboards, not all keys produce the correct sign.

### ISSUE THREAD

**Issue by rabejens** [midwan/amiberry#90]
I am using a German keyboard with Amiberry. Some keys like the number sign or the less/greater sign don't work correctly. I know the scan codes of AmigaOS and PC keyboards are different, but is any custom mapping for *keyboard* keys (not joysticks) planned or even implemented?

**Comment by midwan** [comment]
@rabejens 
Custom Mapping is currently being rewritten to allow mapping of ANY key to the functions available. It will start with allowing mapping of keyboard keys first, then proceed with controller buttons.

But I suspect that is not what you are looking for? Custom Mapping is for mapping controller actions (specifically, CD32 controller) to keys on the keyboard or buttons on a controller.

If your keys don't work as expected inside the emulation, for example on a shell prompt or a text editor, then custom mapping will not help there. What we'll need is multi-lingual support for keyboards, which is not implemented yet.

WinUAE has something like this for a few languages, including German. I believe we could get some ideas from there, unless we come up with a smarter solution. ;)

**Comment by rabejens** [comment]
> If your keys don't work as expected inside the emulation, for example on a shell prompt or a text editor, then custom mapping will not help there.

This is exactly what I'm talking about. I can't enter characters like # in a Shell using a German keyboard. E-UAE had the option to specify something like `kbd_lang=de` in the configuration and it remapped some keys to work correctly, but I don't know how hard this is to port over to Amiberry.

I haven't yet had a more in-depth look at the source code, but I suppose UAE4ARM reads raw scan codes from the keyboard somehow, I can think of a keyboard mapping helper where you press a key or a combination of keys and assign the appropriate Amiga scan codes to this combination. This would mean you have to remap every single key once (the default mapping could be used if no explicit mapping is given).

Some of the internatinal keycodes are already present in `keyboard.h`:

```c
/* The following have different mappings on national keyboards */

#define AK_LBRACKET 0x1A
#define AK_RBRACKET 0x1B
#define AK_SEMICOLON 0x29
#define AK_COMMA 0x38
#define AK_PERIOD 0x39
#define AK_SLASH 0x3A
#define AK_BACKSLASH 0x0D
#define AK_QUOTE 0x2A
#define AK_NUMBERSIGN 0x2B
#define AK_LTGT 0x30
#define AK_BACKQUOTE 0x00
#define AK_MINUS 0x0B
#define AK_EQUAL 0x0C
```

The [AmigaOS Wiki](http://wiki.amigaos.net/wiki/Keymap_Library) has some useful information on this, and [here is a schematic drawing of all Amiga Scan codes.](http://wiki.amigaos.net/w/images/thumb/1/12/LibFig34-2.png/800px-LibFig34-2.png) The problematic keys are `0x2B` and `0x30` which are not present on US keyboards, but `0x2B` is the number sign and the apostroph and `0x30` is the key for less/greater signs.

**Comment by rabejens** [comment]
I forked the project and created a quick and dirty hack that at least makes German keyboards work as well as possible. Some compromises had to be made:

 - The backslash key DID work right on US keyboards, this is now dead (on a real Amiga, the "hash" key was also dead when the keymap was set to usa); when the keymap is set to de, using a US keyboard, the backslash key produces a hash (and a circumflex when shifted)
 - The backslash key is now F11

This is in no means a completed development, it is just meant as a starting point. My idea would be to somehow detect which keyboard is set in the underlying host system and if it is not US-American, but this hack in-place.

It is in the [sdl2-intlkbd branch](https://github.com/rabejens/amiberry/tree/sdl2-intlkbd) of the fork if one wants to try it.

**Comment by Voljega** [comment]
on the same subject, I'm french and using an azerty keyboard and noticed the following :

- some games ask to press '1' '2' '3' or so at the beginning and on a french keyboard for some of them none of the followin keys work : upper '1', shift + upper '1', numerical keypad '1'
if you switch the keyboard type in the linux distribution to 'us/en' to us all games then work by either pressing upper '1' or numerical keypad '1' but some of them only work with numerical keypad

- for some games which ask enter/return to be pressed, only numerical keypad enter/return will work

basically i'm forced to set the keyboard to 'us/en' if I want all games to work even if I have an azerty keyboard

(again i'm on sdl1 version, amiberry 2.1)

**Comment by kcbrown** [comment]
FS-UAE apparently already supports complete key remapping: https://fs-uae.net/input-mapping and http://fs-uae.net/input-actions.

Is that not already incorporated here?  If not, would it not be better to synchronize the two projects in this regard?

**Comment by midwan** [comment]
Note: we should look into this again, now that the input section has been reworked (especially in SDL2)

**Comment by midwan** [comment]
Could anyone please test the situation with the latest release, and see if things have improved now?

**Comment by rabejens** [comment]
I'm currently in the progress of setting up an image with Amiberry and giving it a shot.

**Comment by midwan** [comment]
@rabejens 
Thanks for the feedback! Which distro did you pick for this?
It would help if I can run similar tests from my side on the same once you got it ready.

**Comment by rabejens** [comment]
I want to try Retropie first, on Facebook they said I could install a recent version using the setup script.

**Comment by midwan** [comment]
@rabejens 
Yes, that's correct. If you update the RetroPie-Setup script you can get the latest from source (master branch)

**Comment by rabejens** [comment]
Does this pull the version from your repository so we are talking about the same one, and which one does it use? SDL1 or SDL2?

**Comment by midwan** [comment]
@rabejens 
It pulls the master branch from the repo, and on RetroPie we use the SDL1 version (because it's still based on Jessie).

**Comment by rabejens** [comment]
Then I'll try that first, and after that, get Amibian and compile the SDL2 version from source.

**Comment by rabejens** [comment]
Addendum: What about SDL2 on Retropie? Do I still have to compile it from source there or can I use the one that is included there? I remember earlier versions of Amiberry used SDL2 from Retropie.

**Comment by midwan** [comment]
@rabejens 
We used the SDL2 packages from RetroPie on other distros in the past, to make it easier to compile, but since RetroPie is still running Jessie and using the Legacy driver, using Amiberry with SDL2 + Dispmanx on it is problematic.

We run into the following issues:
- Jessie comes with an older version of libpng (v1.2 vs 1.6 on Stretch) and SDL2 has a requirement on the newer version. At least the newer versions of SDL2, which you want to get the bugfixes.
- The Legacy driver can only use the "rpi" backend of SDL2, but that conflicts with using Dispmanx in the same project. Amiberry uses Dispmanx for the emulation screen on the RPI platform for the best performance.

**Comment by rabejens** [comment]
Then I think I'll go with Amibian right from the start as I reckon this being the better choice for an Amiga emulation machine.

**Comment by midwan** [comment]
@rabejens 
Last I checked Amibian was also on Jessie however, I'm not sure if Gunnar managed to upgrade it and publish a new image yet... ;-)
You can also test on DietPi and a stock Raspbian Stretch.

**Comment by rabejens** [comment]
Ah OK. Yes I remember Amibian being on Jessie, too.

**Comment by rabejens** [comment]
First progress: Using DietPI, I compiled SDL2 from source. I ran into two dependencies not listed [on the Wiki page](https://github.com/midwan/amiberry/wiki/Compile-SDL2-from-source):

```bash
# apt install libfreetype6-dev libgl1-mesa-dev
```

Now, Amiberry is building.

**Comment by midwan** [comment]
@rabejens 
Thanks, those were not listed because the tests are usually made on a stock Raspbian system, which includes them. I'll add them to the Wiki page.

**Comment by rabejens** [comment]
DietPi is also missing bcm_host.h and I don't know how to get it. Trying to update the firmware from dietpi-config seems to just get stuck. Trying the official Raspbian now.

**Comment by rabejens** [comment]
Okay, so far I compiled everything, but I get "SDL_Error: No available video device".

I have no X installed, and I want it to run without X. I followed all steps in the Wiki. What am I missing?

**Comment by rabejens** [comment]
OK, first impressions:

- The special characters still can't be correctly input on a German keyboard - pressing the "#" key still yields the backslash
- In the configuration dialog, it's still not possible to enter capital letters - press Shift-S, get "s"
- The combobox for the disk selection for DF0 and DF1 shows the wrong disk. Insert, e.g., "Install 3.1" in DF0 and "Locale" in DF1, the combobox for DF0 shows "Locale" too although the correct disk IS in DF0

**Comment by Voljega** [comment]
Hello sorry, I am not able to test at the moment, I don't really know ho to compile, and I could try it, but as my platform is Recalbox, I'm not sure I wouldn't be missing any new library

**Comment by midwan** [comment]
@rabejens 
Thanks for the feedback.
Regarding what you mentioned:
- The shift problem in the GUI was resolved a few months ago, when I updated the guisan project contained in the repo. What steps did you take to recreate this? I cannot see the problem in the current version (either `master` or `dev` branch).
- The dropdown bug was fixed in an earlier commit, in the `dev` branch. It will be merged back into master with the next release.

That leaves us the `#` key on a German keyboard. I'll see what I can do about that.

**Comment by midwan** [comment]
@rabejens 
Please test again with the latest `dev` branch changes:
The Insert key should now map to the `#` key correctly, and the `>/<` key should get correctly detected on a keyboard that has it as well.

I tested it using SysTest.adf (from Aminet) and it seems to work as expected.
